### PR TITLE
FHIRPath Patch

### PR DESCRIPTION
**Pull request by mikaelweave**
## Description
Implements FHIRPath Patch as defined on [the HL7 website](http://hl7.org/fhir/fhirpatch.html). Co-coded with @mattpasco-microsoft and @stgricci.

##  Done
- [x] Implemented FHIRPatch
- [x] Check edge cases with FHIR chat group. Document behavior on undefined areas.
- [x] Re-use extensions and helper methods in `fhir-server` repo. (no `ElementNode` extensions exist elsewhere)
- [x] Complete unit testing.
- [x] Add E2E testing.
- [x] Fix FHIRPatch in bundles.
- [x] Add example `.http` file for FHIRPatch.

## Related issues
Addresses #1268. Related to unmerged PR #1272 by @valeneiko.

## FHIR Team Checklist
- [ ] **Update the title** of the PR to be succinct and less than 50 characters
- [ ] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [ ] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [ ] Tag the PR with Azure API for FHIR if this will release to the Azure API for FHIR managed service (CosmosDB or common code related to service)
- [ ] Tag the PR with Azure Healthcare APIs if this will release to the Azure Healthcare APIs managed service (Sql server or common code related to service)
- [ ] CI is green before merge
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Core/Messages/Patch/ConditionalFhirPatchResourceRequest.cs
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Core/Messages/Patch/ConditionalFhirPatchResourceRequest.cs:16]: >ConditionalFhirPatchResourceRequest [](http://example.com/codeflow?start=24&length=35)

This and FhirPatchResourceRequest looks excessive for me.
What are they doing what ConditionalPatchResourceRequest can't do?

File: src/Microsoft.Health.Fhir.Shared.Api/Controllers/FhirController.cs
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Shared.Api/Controllers/FhirController.cs:482]: >ConditionalFhirPatchResourceRequest [](http://example.com/codeflow?start=20&length=35)

This seems excessive for me. That whole class does same thing as ConditionalPatchRequest, with only difference is what is payload it carries.

I would do following:


Either make `PatchResourceRequest` generic by payload (either Parameters or JsonPatchDocument or just have two properites inside and two different constructors. 

Or make `PatchResourceRequest` an abstract class and derive two classes out of it FhirJson and Json

Rename JsonPatchService into PatchService and let it have either two public patch methods, or one public patch which would accept ConditionalPatchRequest and two private methods to handle patching. 

It feels like too much same code right now.

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/FhirPathPatch/FhirPathPatchBuilder.cs
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/FhirPathPatch/FhirPathPatchBuilder.cs:48]: >Apply [](http://example.com/codeflow?start=24&length=5)

Since we use Firely maybe it make sense to check out their patch service implementation: [https://github.com/FirelyTeam/spark/blob/r4/master/src/Spark.Engine/Service/FhirServiceExtensions/PatchService.cs](https://github.com/FirelyTeam/spark/blob/r4/master/src/Spark.Engine/Service/FhirServiceExtensions/PatchService.cs)

File: src/Microsoft.Health.Fhir.Core/Messages/Patch/ConditionalFhirPatchResourceRequest.cs
Comment (mikaelweave) [reply src/Microsoft.Health.Fhir.Core/Messages/Patch/ConditionalFhirPatchResourceRequest.cs:16]: @mattpasco-microsoft can you comment on this one?

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/FhirPathPatch/FhirPathPatchBuilder.cs
Comment (Ivanidzo4ka) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/FhirPathPatch/FhirPathPatchBuilder.cs:48]: My intent in comment was to make sure we don't reinvent bike, and it's good to know what you check other implementation.

Comment (mikaelweave) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/FhirPathPatch/FhirPathPatchBuilder.cs:48]: @Ivanidzo4ka - did some digging. Spark's implementation doesn't pass any tests involving [Anonymous Types](https://github.com/FHIR/fhir-test-cases/blob/752b01313ecbc1e13a942e1b3e25c96b3f7f3449/r5/patch/fhir-path-tests.xml#L450) or [Nested Primitives](https://github.com/FHIR/fhir-test-cases/blob/752b01313ecbc1e13a942e1b3e25c96b3f7f3449/r5/patch/fhir-path-tests.xml#L297). I remember trying to fix Spark's PatchService code, but failing since the code is complex. I'm going to do a simple performance comparison later just to make sure the two approaches aren't wildly different. As of right now, the approach in this PR is superior because it meets all of the FHIRPatch tests defined by HL7.

Comment (mikaelweave) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/FhirPathPatch/FhirPathPatchBuilder.cs:48]: @Ivanidzo4ka - performance difference is negligible testing with 100 sequential FHIRPatch `add` operations and 100 sequential FHIRPatch `replace` operations.

File: src/Microsoft.Health.Fhir.Core/Messages/Patch/ConditionalFhirPatchResourceRequest.cs
Comment (mattpasco-microsoft) [reply src/Microsoft.Health.Fhir.Core/Messages/Patch/ConditionalFhirPatchResourceRequest.cs]: Agreed, pushed some changes to reduce complexity so the only duplication is in the patch service itself

File: src/Microsoft.Health.Fhir.Shared.Api/Controllers/FhirController.cs
Comment (mattpasco-microsoft) [reply src/Microsoft.Health.Fhir.Shared.Api/Controllers/FhirController.cs]: I templated this and reduced complexity

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/ConditionalPatchResourceHandler.cs
Comment (mattpasco-microsoft) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/ConditionalPatchResourceHandler.cs:45]: @Ivanidzo4ka @mikaelweave I templated away both the ResourceRequest and ConditionalPatchResourceRequest. Ideally we could guard the TData type to be either Parameters or JsonPatchDocument but I foresee defering type check logic somewhere else. I really hate this code for picking the service but im not sure a full factory pattern is necessary. My other idea would be to lazy an instance of the JsonPatchService and FhirPatchService.

Would love some more input here

Comment (mikaelweave) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/ConditionalPatchResourceHandler.cs]: @mattpasco-microsoft this works but the resource handlers need to be registered explicitly with the DI container due to MediatR and possibly even .NET. [See here for details](https://github.com/jbogard/MediatR/issues/674).

I added the registration for the Patch handlers and used the container for`JsonPatchService` and `FhirParametersPatchService`. I added them where it made sense to me but should be reviewed by someone for familiar with the repo (@Ivanidzo4ka and/or @brendankowitz).

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/BasePatchService.cs
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/BasePatchService.cs:28]: >protected abstract Resource GetPatchedJsonResource(FhirJsonNode node, T operations); [](http://example.com/codeflow?start=8&length=84)

nit: I don't think this been used outside of internal implementation of each patch service, so I would remove it from base implementation and keep each method as private inside specific implementations.

Comment (mattpasco-microsoft) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/BasePatchService.cs]: Agreed, maybe there is some further refactoring we could do to move shared code.

**Comment by mikaelweave** [comment]
@Ivanidzo4ka are there any docs or E2E test that we need to add for this PR? Besides those two items, this is about ready.

**Comment by Ivanidzo4ka** [comment]
@mikaelweave `docs` can mean two things in my mind. One is for developers who work with this repository. Second is for our azure customers who use this server as a product, if you mean second one I would suggest you to talk to @CaitlinV39 since I don't have much expertise, and it probably would involve commit to different repository. If you thinking about first, I think doc/rest is bare minimum, and you can add more content into doc folder regarding how patch works in general, but something tells me you pursue second type of docs...

For E2E tests feel free to add more tests to this file:
https://github.com/microsoft/fhir-server/blob/main/test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/PatchTests.cs 
Note that it uses client class https://github.com/microsoft/fhir-server/blob/6dfd829796954236bec67c8c71e98bbdba5bf6c3/src/Microsoft.Health.Fhir.Shared.Client/FhirClient.cs#L237 which currently hardcodes type of the content, so you will need to touch it as well (maybe by having two methods JsonPatch* and ParametersPatch* instead of Patch*)

**Comment by CaitlinV39** [comment]
> @mikaelweave `docs` can mean two things in my mind. One is for developers who work with this repository. Second is for our azure customers who use this server as a product, if you mean second one I would suggest you to talk to @CaitlinV39 since I don't have much expertise, and it probably would involve commit to different repository. If you thinking about first, I think doc/rest is bare minimum, and you can add more content into doc folder regarding how patch works in general, but something tells me you pursue second type of docs...
> 
> For E2E tests feel free to add more tests to this file: https://github.com/microsoft/fhir-server/blob/main/test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/PatchTests.cs Note that it uses client class
> 
> https://github.com/microsoft/fhir-server/blob/6dfd829796954236bec67c8c71e98bbdba5bf6c3/src/Microsoft.Health.Fhir.Shared.Client/FhirClient.cs#L237
> 
> which currently hardcodes type of the content, so you will need to touch it as well (maybe by having two methods JsonPatch* and ParametersPatch* instead of Patch*)

@mikaelweave - We would love if you could add a doc/rest that gives users examples of how to use the FHIRPath Patch. Once this is merged, I'll work to get Azure Docs updated with the details and ask for your review/input on things I don't know :) Typically the rest file helps me do this without having to hassle you as much!

**Comment by mikaelweave** [comment]
> > @mikaelweave `docs` can mean two things in my mind. One is for developers who work with this repository. Second is for our azure customers who use this server as a product, if you mean second one I would suggest you to talk to @CaitlinV39 since I don't have much expertise, and it probably would involve commit to different repository. If you thinking about first, I think doc/rest is bare minimum, and you can add more content into doc folder regarding how patch works in general, but something tells me you pursue second type of docs...
> > For E2E tests feel free to add more tests to this file: https://github.com/microsoft/fhir-server/blob/main/test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/PatchTests.cs Note that it uses client class
> > https://github.com/microsoft/fhir-server/blob/6dfd829796954236bec67c8c71e98bbdba5bf6c3/src/Microsoft.Health.Fhir.Shared.Client/FhirClient.cs#L237
> > 
> > which currently hardcodes type of the content, so you will need to touch it as well (maybe by having two methods JsonPatch* and ParametersPatch* instead of Patch*)
> 
> @mikaelweave - We would love if you could add a doc/rest that gives users examples of how to use the FHIRPath Patch. Once this is merged, I'll work to get Azure Docs updated with the details and ask for your review/input on things I don't know :) Typically the rest file helps me do this without having to hassle you as much!

Thanks for the info! We have a `.http` file in the docs folder! While re-testing everything, I did see that FHIRPatch in bundles errors because the correct patch type cannot be resolved. I'm going to move this back to draft until this and E2E testing gets resolved. I think E2E testing is needed since the patch type selector code is non-trivial.

I'm not working for the next 2 weeks, but can look once I'm back. @mattpasco-microsoft, if you have time could you look?

**Comment by mikaelweave** [comment]
@Ivanidzo4ka E2E tests added, code is done. Ready for final review/testing.

CC @mattpasco-microsoft

**Comment by mikaelweave** [comment]
@LTA-Thinking I can look today! Thanks for running the pipeline. 👍🏽

**Comment by Ivanidzo4ka** [comment]
Failure are caused by making Handler (ConditionalPatch and just Patch) generic, and code which we use in tests, doesn't know how to create object for generic types (which makes sense.).

Sorry for diverting you on this path, and not thinking through all of it.

I'm ok with reverting code back to previous state without generics. As another option, instead of having TData as parameter we can introduce abstract class PatchPayload and combine it with `BasePatchService` by having `JsonPatchPayload` which knows how to patch  application/json and `ParametersPatchPayload` which knows how to patch application/parameters.

**Comment by mikaelweave** [comment]
@Ivanidzo4ka no problem, good learning experience. I wasn't in love with this approach (especially changes to [MediationModule.cs](src/Microsoft.Health.Fhir.Shared.Api/Modules/MediationModule.cs)). I'll look at the abstract approach to reduce duplicated code. If it gets too complex, I'll just duplicate the code for easier understandability in the future.

**Comment by mikaelweave** [comment]
@Ivanidzo4ka sorry for the delay, dealt with computer issues this week. Question about your approach - would `JsonPatchPayload` and `ParametersPatchPayload` be sent by the `PatchResourceRequest` function? Or would they be compiled in the handler?

After looking at this, I could easily make `PatchResourceRequest` and `PatchResourceHandler` abstract and then send different requests to different handlers with minimal code duplication. This would continue the MediatR pattern.

**Comment by Ivanidzo4ka** [comment]
How I saw it in my head:
You have controller and controller creates handler request with either `PatchResourcePayload` or `ParameterResourcePayload`.

We have one handler which handles request by fetching document from DB, and asking Payload class to patch it, get result back and validate it.

**Comment by mikaelweave** [comment]
I see what you are saying now, @Ivanidzo4ka. I changed the code but have one more question before final testing. I placed the `PatchPayload` files in `Microsoft.Health.Fhir.Shared.Core`. I cannot reference them from `Microsoft.Health.Fhir.Core` because adding a reference to Shared from that project creates a dependency loop. Have you seen this issue before in the fhir-server codebase?

Some possible solutions that come to mind:
- Define an interface in another assembly that is implemented in `Microsoft.Health.Fhir.Shared.Core`
- Define the classes in the `Microsoft.Health.Fhir.Core` and use extensions for the `PatchPayload` logic that should be in the Shared project
- Assemble the payload in Shared and use separate Requests to know which to build

**Comment by mikaelweave** [comment]
Thanks for the help @LTA-Thinking, this is the largest project I've worked on and want to make sure I stick with the established patterns.

**Comment by Ivanidzo4ka** [comment]
Let's talk first about difference between `Shared.Core` and `Core` and why we even have two projects.

Idea of having `Core` project should be transparent, I hope, bunch of basic classes we can use in different parts of solution. Very basic things applicable for anything in project.

Why we have `Shared.Core`? Well, we work with FHIR spec, and sometimes we need to work with specific entities defined in FHIR spec. As usual, let's pick Patient entity. There is 3 different variations of FHIR spec we work with: STU3, R4, R5. In all three of them Patient entity exist. But in some of them it has certain properties, and in some it doesn't. Spec evolves and changes over time, and STU3, R4,R5 is just snapshots of the spec in certain time.

So I want to have class which has reference on Patient class, how I do it? We do it by having 3 different projects (STU3, R4, and R5) and having identical classes with reference to Patient class in them. Copying code around 3 different projects, is cumbersome and error prone task.

To make our life easier, we have `Shared.Core` project which has code for all 3 projects, and each project just reuses code but for each project meaning of Patient class is different. Basically `Shared.Core` is collection of templates or generics which we resolve during compilation into specific classes.

Wherefore `Shared.Core` is imbued with domain knowledge, while classes in `Core` are basic things which have no knowledge of FHIR spec.

You put `*Payload` classes in `Shared.Core` because they have references on FHIR Spec classes I assume (Parameters`).
And you define Handler request in core, because it's basic class, untouched by FHIR spec specifics.
But you also have problem of putting `*Payload` into request. This is typical inversion of dependency problem.

I personally would go with abstract class `PatchPayload` in `Core` project which would have ` ResourceElement Patch(ResourceWrapper resourceToPatch, WeakETag weakETag)` method. 

You can have concreate implementations of this class (Parameters, Json) in `Shared.Core` project.
`FhirController` I think would be the place where you have to create concreate implementations, and `FhirController` already knows about specific FHIR spec, so shouldn't be a problem. (`FhirController` lives in another `Shared` -> `Shared.Web` project)

I personally would choose abstract class over interface, because I don't expect someone else have functionality of payload. But if you find it's easier for you to implement that way, go for it.

**Comment by mikaelweave** [comment]
Thanks for the detailed explanation @Ivanidzo4ka. This clears up a LOT on the repository. The code reuse between STU3, R4, and R5 was obvious. Beyond that, I didn't understand the difference between `Core` and `Shared`.  I saw the separation but didn't understand the *why* behind the separation. I distinctly remember being unsure about where to put the Patch library initially and think I got lucky.

I'm also not super strong in design patterns, so keep that in mind while forming your persona of how a developer in my position understands the repo. I know what they are, but until you get to use them a lot you tend to miss things. 😄  

One piece of feedback (I can create an issue after the PR) is I looked for Developer docs and couldn't find any in the root of the repo or the `docs` folder. The sequence diagrams were helpful, but some architectural guidance in a `DEVELOPERS.md` file would help outside contributors.

❓  Final question, I put `PatchPayload in Core.Features.Operations.Resources` to match the other projects and keep everything in the same namespace. Any issues here? Not sure if it's super important but it made sense to me. This required the updating of some string definitions from starting with `Resources` to `Core.Resources`.

Seriously, THANK YOU for the long comment explaining your reasoning. I have become a better developer thanks to the time you've put into that comment and this PR.

**Comment by LTA-Thinking** [comment]
@CaitlinV39 Adding developer focused documentation would be helpful for our open source contributors, and may encourage more contributions like this.

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/FhirPathPatch/FhirPathPatchBuilder.cs
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/FhirPathPatch/FhirPathPatchBuilder.cs:12]: Just a small one, namespaces don't look consistent with the rest of the project

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/FhirPathPatch/Operations/EOperationType.cs
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/FhirPathPatch/Operations/EOperationType.cs:11]: Perhaps `PatchOperationType` for consistency?

File: src/Microsoft.Health.Fhir.Core/Messages/Patch/ConditionalPatchResourceRequest.cs
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Messages/Patch/ConditionalPatchResourceRequest.cs:32]: I think this is good, always felt a little strange having the asp.net JsonPatchDocument here directly :)

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/FhirPathPatch/FhirPathPatchBuilder.cs
Comment (mikaelweave) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/FhirPathPatch/FhirPathPatchBuilder.cs]: Good catch, thanks. We wrote this as a standalone library and never updated this. I've combed the logic piece of FHIRPatch and cleanup up a bunch of things.

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/FhirPathPatch/Operations/EOperationType.cs
Comment (mikaelweave) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/FhirPathPatch/Operations/EOperationType.cs]: Done, makes more sense. I didn't prefix `FhirPathPatch` on any of the Operation classes because all the code in `FhirPathPatch` is internal except for the builder and a helper used for testing.

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/FhirPathPatch/Operations/IOperation.cs
Comment (mikaelweave) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/FhirPathPatch/Operations/IOperation.cs]: Removed the interface, it wasn't giving any benefit. No code outside `FhirPathPatch' needs to know the operation details.

File: src/Microsoft.Health.Fhir.Core/Messages/Patch/ConditionalPatchResourceRequest.cs
Comment (mikaelweave) [reply src/Microsoft.Health.Fhir.Core/Messages/Patch/ConditionalPatchResourceRequest.cs:32]: Agreed! Big thanks to Ivan for the help and contributions to the PR by pushing us in the right direction.

File: docs/rest/FhirPatchRequests.http
Comment (LTA-Thinking) [inline docs/rest/FhirPatchRequests.http:99]: Super nitpick: I think this example should be above the example that removes a phone number to allow the remove example to demonstrate the functionality of the condition in the valueString. Right now there is only one phone number so it is unclear if the conditional statement is working.

Comment (mikaelweave) [reply docs/rest/FhirPatchRequests.http]: Good callout. I'm going to run all the rest requests again in a couple of hours for a sanity check.

Comment (mikaelweave) [reply docs/rest/FhirPatchRequests.http]: All looks good on these @LTA-Thinking. Tested and still working as intended.

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/Patch/FhirPatchDeleteTests.cs
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/Patch/FhirPatchDeleteTests.cs:124]: Nit: Typo in the test name

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/Patch/FhirPatchInsertTests.cs
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/Patch/FhirPatchInsertTests.cs:124]: Nit: I think the test name is wrong. It looks like this test is for an index that is too high, not an uninitialized list.

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/ConditionalPatchTests.cs
Comment (LTA-Thinking) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/ConditionalPatchTests.cs:91]: For this test the JSON and FHIR patches should be in separate tests. Since they are working on the same patient if the JSON one succeeds then there are no unique changes made by the FHIR patch to check that it worked properly.

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/ConditionalPatchTests.cs
Comment (LTA-Thinking) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/ConditionalPatchTests.cs:155]: Same comment as above. When changes are being made there should be separate tests.

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/FhirPathPatchTests.cs
Comment (LTA-Thinking) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/FhirPathPatchTests.cs:55]: Nit: Typo in method name.

**Review by LTA-Thinking — COMMENTED** [review]
Finished with the test code. These were the last files for me to look at. We should be good to go after these changes are in. Thanks for working with us on this change, it is a big one.

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/Patch/FhirPatchInsertTests.cs
Comment (mikaelweave) [reply src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/Patch/FhirPatchInsertTests.cs:124]: Test name was right, content misleading/wrong. Changed insert index to 0 to actually ensure the error we are getting matches the intended test case.

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/ConditionalPatchTests.cs
Comment (mikaelweave) [reply test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/ConditionalPatchTests.cs:103]: Agreed. Was trying to be cute to remove duplication of code but it didn't really get us much. I separated these out.

**Comment by mikaelweave** [comment]
@LTA-Thinking Thanks for the careful eye, I definitely left off some items in my final check. I'll go through the whole PR again just to be sure.

Happy to work with you all on this to produce high quality code I can be proud of. I hope to commit more in the future if I see other opportunities to enable use cases. 👍🏽

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Resources/Patch/PatchPayload.cs
Comment (mikaelweave) [inline src/Microsoft.Health.Fhir.Core/Features/Operations/Resources/Patch/PatchPayload.cs:38]: @LTA-Thinking - here I went from an `IModelInfoProvider` provided by the DI container to this statically provided version. Any issues with this approach?

I changed to this approach as `IModelInfoProvider` isn't injected in `FhirController`.

Comment (LTA-Thinking) [reply src/Microsoft.Health.Fhir.Core/Features/Operations/Resources/Patch/PatchPayload.cs:38]: This is the right way to do it. We use this pattern in other parts of our code.

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/ConditionalPatchTests.cs
Comment (LTA-Thinking) [reply test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/ConditionalPatchTests.cs:187]: This one still needs to be split into two tests.

Comment (mikaelweave) [reply test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/ConditionalPatchTests.cs:203]: I'm going to blame this on yesterday being Monday and point at the cleanup commit I did yesterday to distract you. 😆 

But seriously, missed this one, it's now split into multiple tests. Thanks.

**Review by LTA-Thinking — APPROVED** [review]

**Comment by LTA-Thinking** [comment]
Thanks @mikaelweave and @mattpasco-microsoft!

**Pull request merged**
